dbms/src/Interpreters/RowRefs.h

```diff
@@ -144,48 +144,52 @@ class SortedLookupVector
         array.push_back(std::forward<U>(x), std::forward<TAllocatorParams>(allocator_params)...);
     }
 
-    // Transition into second stage, ensures that the vector is sorted
-    typename Base::const_iterator upper_bound(const TEntry & k)
+    const RowRef * upperBound(const TEntry & k, bool ascending)
     {
-        sort();
-        return std::upper_bound(array.cbegin(), array.cend(), k);
+        sort(ascending);
+        auto it = std::upper_bound(array.cbegin(), array.cend(), k, (ascending ? less<true> : less<false>));
+        if (it != array.cend())
+            return &(it->row_ref);
+        return nullptr;
     }
 
-    // After ensuring that the vector is sorted by calling a lookup these are safe to call
-    typename Base::const_iterator cbegin() const { return array.cbegin(); }
-    typename Base::const_iterator cend() const { return array.cend(); }
+    const RowRef * lowerBound(const TEntry & k, bool ascending)
+    {
+        sort(ascending);
+        auto it = std::lower_bound(array.cbegin(), array.cend(), k, (ascending ? less<true> : less<false>));
+        if (it != array.cend())
+            return &(it->row_ref);
+        return nullptr;
+    }
 
 private:
     std::atomic<bool> sorted = false;
     Base array;
     mutable std::mutex lock;
 
-    struct RadixSortTraits : RadixSortNumTraits<TKey>
+    template <bool ascending>
+    static bool less(const TEntry & a, const TEntry & b)
```

Comment:
'Descending less' looks like an oxymoron :)
I guess we can just have two comparators, something like asof_asc() and asof_desc().